README.md
[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/0.1.0/active.svg)](http://www.repostatus.org/#active)
[![Is the package on CRAN?](http://www.r-pkg.org/badges/version/assertive.datetimes)](http://www.r-pkg.org/pkg/assertive.datetimes)
[![SemaphoreCI Build Status](https://semaphoreci.com/api/v1/projects/8d10dcae-f422-4e66-8cad-43e510ccce5c/635114/badge.svg)](https://semaphoreci.com/richierocks/assertive-datetimes)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/jkk8d4rdi6h2hnkh?svg=true)](https://ci.appveyor.com/project/richierocks/assertive-datetimes)
[![Research software impact](http://depsy.org/api/package/cran/assertive.datetimes/badge.svg)](http://depsy.org/package/r/assertive.datetimes)

# assertive.datetimes

A set of predicates and assertions for checking the properties of dates and times.  Most of the documentation is on the *[assertive](https://bitbucket.org/richierocks/assertive)* page.  End-users will usually want to use *assertive* directly.


### Installation

To install the stable version, type:

```{r}
install.packages("assertive.datetimes")
```

To install the development version, you first need the *devtools* package.

```{r}
install.packages("devtools")
```

Then you can install the *assertive.datetimes* package using

```{r}
library(devtools)
install_bitbucket("richierocks/assertive.datetimes")
```

### Predicates

`is_date_string` checks character vectors to see if they contain dates and times.

`is_before`, and `is_after` check whether `Date`s or `POSIXt` objects are before or after some time.

`is_in_past` and `is_in_future` are like the above, compared to the present time.

### Assertions

All predicates return a vector and have two corresponding assertions.  For example,
`is_after` has `assert_all_are_after` and `assert_any_are_after`.

NEWS
0.0-3 Fns comparing times will always coerce to POSIXct before making comparison.
0.0-2 Maintenance release updating tests to match new assertive.base behaviour.
0.0-1 Content extracted from assertive 0.3-0, and tidied.


tests/testthat.R
library(testthat)
library(devtools)
library(assertive.datetimes)

with_envvar(
  c(LANG = "en_US"),
  test_check("assertive.datetimes")
)


tests/testthat/test-is-date-string.R
test_that(
  "test.is_date_string.a_character_vector.returns_true_when_string_contains_a_date", 
  {
    x <- c("1999-12-31 23:59:59", "1979-08-01 01:00:00", "31 Dec 1999 11:59:59PM", 
           "not a date", "NA")
    expected <- c(TRUE, TRUE, FALSE, FALSE, FALSE)
    expect_equal(strip_attributes(actual <- is_date_string(x)), expected)
    expect_equal(names(actual), unname(x))
    expect_equal(
      cause(actual),
      noquote(rep.int(c("", "bad format"), c(2, 3)))
    )
  }
)


R/is-time.R
#' @include workaround.R
#' @rdname is_in_past   
#' @export
is_after <- function(x, y, .xname = get_name_in_parent(x), 
  .yname = get_name_in_parent(y))
{
  x <- coerce_to(x, "POSIXct", .xname)
  y <- coerce_to(y, "POSIXct", .yname)
  call_and_name_retro(
    function(x)
    {
      ok <- x > y
      set_cause(ok, "before")
    },
    x
  )
}

#' @rdname is_in_past   
#' @export
is_before <- function(x, y, .xname = get_name_in_parent(x), 
  .yname = get_name_in_parent(y))
{  
  x <- coerce_to(x, "POSIXct", .xname)
  y <- coerce_to(y, "POSIXct", .yname)
  call_and_name_retro(
    function(x)
    {
      ok <- x < y
      set_cause(ok, "after")
    },
    x
  )
}

#' @rdname is_in_past   
#' @export
is_in_future <- function(x, .xname = get_name_in_parent(x))
{
  x <- coerce_to(x, "POSIXct", .xname)
  call_and_name_retro(
    function(x)
    {
      ok <- x > Sys.time()
      set_cause(ok, "in past")
    },
    x
  )
}

#' Is the input in the past/future?
#'
#' Checks to see if the input is a time in the past/future, or before/after
#' some time point.
#'
#' @param x \code{Date} or \code{POSIXt} input to check.
#' @param y Another date-time object to compare against.
#' @param na_ignore A logical value.  If \code{FALSE}, \code{NA} values
#' cause an error; otherwise they do not.  Like \code{na.rm} in many
#' stats package functions, except that the position of the failing
#' values does not change.
#' @param severity How severe should the consequences of the assertion be?  
#' Either \code{"stop"}, \code{"warning"}, \code{"message"}, or \code{"none"}.
#' @param .xname Not intended to be used directly.
#' @param .yname Not intended to be used directly.
#' @return The \code{is_*} function return \code{TRUE} if the input is 
#' a time in the future/past.  The \code{assert_*} functions return nothing but
#' throw an error if the corresponding \code{is_*} function returns 
#' \code{FALSE}.
#' @details The current time is determined by \code{Sys.time}, and the 
#' input is coerced to \code{POSIXct} format.
#' @note Note that the print method for \code{POSIXct} objects means that the
#' cause attribute (in the event of failures) is not shown.  You can still 
#' access it via, e.g., \code{cause(is_in_past(x))}.
#' @seealso \code{\link{Sys.time}}.
#' @examples
#' datetime <- Sys.time() + c(-1, 100)
#' is_in_past(datetime)
#' is_in_future(datetime)
#' date <- Sys.Date() + c(-1, 100)
#' 
#' # more generally, compare against any date-time
#' is_before(datetime, as.POSIXct("9999-12-31"))
#' is_after(datetime, as.POSIXct("0001-01-01"))
#' @importFrom assertive.types is_date
#' @export
is_in_past <- function(x, .xname = get_name_in_parent(x))
{
  x <- coerce_to(x, "POSIXct", .xname)
  call_and_name_retro(
    function(x)
    {
      ok <- x < Sys.time()
      set_cause(ok, "in future")
    },
    x
  )
}


R/assert-is-time.R
#' @include imports.R

#' @rdname is_in_past
#' @export
assert_all_are_after <- function(x, y, na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{              
  .xname = get_name_in_parent(x)  
  .yname = get_name_in_parent(y)                                       
  msg <- gettextf("%s are not all after %s.", .xname, .yname)
  assert_engine(
    is_after,
    x, 
    .xname = .xname,
    .yname = .yname,
    msg = msg, 
    na_ignore = na_ignore,
    severity = severity
  ) 
}

#' @rdname is_in_past
#' @export
assert_any_are_after <- function(x, y, na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{                
  .xname = get_name_in_parent(x)  
  .yname = get_name_in_parent(y)                                     
  msg <- gettextf("%s are all before %s.", .xname, .yname)
  assert_engine(
    is_after,
    x, 
    .xname = .xname,
    .yname = .yname,
    msg = msg, 
    what = "any",
    na_ignore = na_ignore,
    severity = severity
  )   
}
 
#' @rdname is_in_past
#' @export
assert_all_are_before <- function(x, y, na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{            
  .xname = get_name_in_parent(x)  
  .yname = get_name_in_parent(y)                                           
  msg <- gettextf("%s are not all before %s.", .xname, .yname)
  assert_engine(
    is_before,
    x, 
    .xname = .xname,
    .yname = .yname,
    msg = msg, 
    na_ignore = na_ignore,
    severity = severity
  ) 
}

#' @rdname is_in_past
#' @export
assert_any_are_before <- function(x, y, na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{            
  .xname = get_name_in_parent(x)  
  .yname = get_name_in_parent(y)                                       
  msg <- gettextf("%s are all after %s.", .xname, , .yname)
  assert_engine(
    is_before,
    x, 
    .xname = .xname,
    .yname = .yname,
    msg = msg, 
    what = "any",
    na_ignore = na_ignore,
    severity = severity
  )   
}
 
#' @rdname is_in_past
#' @export
assert_all_are_in_future <- function(x, na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{                            
  .xname = get_name_in_parent(x)                         
  msg <- gettextf("%s are not all in the future.", .xname)
  assert_engine(
    is_in_future,
    x, 
    .xname = .xname,
    msg = msg, 
    na_ignore = na_ignore,
    severity = severity
  ) 
}

#' @rdname is_in_past
#' @export
assert_any_are_in_future <- function(x, na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{                        
  .xname = get_name_in_parent(x)                              
  msg <- gettextf("%s are all in the past.", .xname)
  assert_engine(
    is_in_future, 
    x, 
    .xname = .xname,
    msg = msg, 
    what = "any",
    na_ignore = na_ignore,
    severity = severity
  )   
}
 
#' @rdname is_in_past
#' @export
assert_all_are_in_past <- function(x, na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{                 
  .xname = get_name_in_parent(x)                                     
  msg <- gettextf("%s are not all in the past.", .xname)
  assert_engine(
    is_in_past, 
    x, 
    .xname = .xname,
    msg = msg, 
    na_ignore = na_ignore,
    severity = severity
  )    
}

#' @rdname is_in_past
#' @export
assert_any_are_in_past <- function(x, na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{                 
  .xname = get_name_in_parent(x)                                     
  msg <- gettextf("%s are all in the future.", .xname)
  assert_engine(
    is_in_past, 
    x, 
    .xname = .xname,
    msg = msg, 
    what = "any",
    na_ignore = na_ignore,
    severity = severity
  )  
}


R/assert-is-date-string.R
#' @rdname is_date_string
#' @export
assert_all_are_date_strings <- function(x, format = "%F %T", na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{    
  .xname <- get_name_in_parent(x)
  msg <- gettextf(
    "%s contains strings that are not in the date format '%s'.", 
    get_name_in_parent(x),
    format, 
    .xname
  )
  assert_engine(
    is_date_string, 
    x, 
    format = format,
    .xname = .xname,
    msg = msg, 
    na_ignore = na_ignore,
    severity = severity
  )
}

#' @rdname is_date_string
#' @export
assert_any_are_date_strings <- function(x, format = "%F %T", na_ignore = FALSE, 
  severity = getOption("assertive.severity", "stop"))
{                   
  .xname <- get_name_in_parent(x)                                  
  msg <- gettextf(
    "%s contains no strings that are in the format '%s'.", 
    get_name_in_parent(x),
    format, 
    .xname
  )
  assert_engine(
    is_date_string, 
    x, 
    format = format,
    .xname = .xname,
    msg = msg, 
    what = "any", 
    na_ignore = na_ignore,
    severity = severity
  )      
}


R/is-date-string.R
#' Does the character vector contain dates? 
#' 
#' Checks that the input contains dates or times.
#' 
#' @param x Input to check.
#' @param format Expected format of the dates.  See 
#' \code{\link[base]{strptime}}.
#' @param na_ignore A logical value.  If \code{FALSE}, \code{NA} values
#' cause an error; otherwise they do not.  Like \code{na.rm} in many
#' stats package functions, except that the position of the failing
#' values does not change.
#' @param severity How severe should the consequences of the assertion be?  
#' Either \code{"stop"}, \code{"warning"}, \code{"message"}, or \code{"none"}.
#' @param .xname Not intended to be used directly.
#' @return A logical vector that is \code{TRUE} when the input contains valid 
#' dates or times.
#' @examples
#' x <- c("9999-12-31 23:59:59", "wednesday", NA)
#' is_date_string(x)
#' assert_all_are_date_strings("01Aug1979", format = "%d%b%Y") #My DOB!
#' @seealso \code{\link[base]{strptime}} for specifying formats, and the 
#' \code{lubridate} package for automatic guessing of date formats (and other 
#' date manipulation functions).
#' @importFrom assertive.base use_first
#' @export
is_date_string <- function(x, format = "%F %T", .xname = get_name_in_parent(x))
{
  x <- coerce_to(x, "character", .xname)
  format <- use_first(format)
  call_and_name(
    function(x) 
    {
      ok <- !is.na(strptime(x, format))
      set_cause(ok, ifelse(is.na(x), "missing", "bad format"))
    }, 
    x
  )  
}

